source/slang/slang-emit-cuda.cpp

```diff
@@ -349,6 +349,150 @@ void CUDASourceEmitter::emitLoopControlDecorationImpl(IRLoopControlDecoration* d
     }
 }
 
+static bool _areEquivalent(IRType* a, IRType* b)
+{
+    if (a == b)
+    {
+        return true;
+    }
+    if (a->op != b->op)
+    {
+        return false;
+    }
+
+    switch (a->op)
+    {
+        case kIROp_VectorType:
+        {
+            IRVectorType* vecA = static_cast<IRVectorType*>(a);
+            IRVectorType* vecB = static_cast<IRVectorType*>(b);
+
+            if (GetIntVal(vecA->getElementCount()) != GetIntVal(vecB->getElementCount()))
+            {
+                return false;
+            }
+            return _areEquivalent(vecA->getElementType(), vecB->getElementType());
+        }
+        case kIROp_MatrixType:
+        {
+            IRMatrixType* matA = static_cast<IRMatrixType*>(a);
+            IRMatrixType* matB = static_cast<IRMatrixType*>(b);
+
+            if (GetIntVal(matA->getColumnCount()) != GetIntVal(matA->getColumnCount()) ||
+                GetIntVal(matA->getRowCount()) != GetIntVal(matA->getRowCount()))
+            {
+                return false;
+            }
+
+            return _areEquivalent(matA->getElementType(), matB->getElementType());
+        }
+        default:
+        {
+            IRBasicType* basicA = as<IRBasicType>(a);
+            IRBasicType* basicB = as<IRBasicType>(b);
+
+            return basicA && basicB;
+        }
+    }
+}
+
+void CUDASourceEmitter::_emitInitArrayValue(IRType* dstType, IRInst* value)
+{
+    // When constructing a matrix or vector from a single value this is already handled by the default path
+
+    switch (value->op)
+    {
+        case kIROp_Construct:
+        case kIROp_MakeMatrix:
+        case kIROp_makeVector:
+        {
+            IRType* type = value->getDataType();
+
+            // If the types are the same, we can can just break down and use
+            if (_areEquivalent(dstType, type))
```

Comment:
Are there cases where this check doesn't pass (that is, where we are initializing something from an expression of the wrong type)? If so, that would seem to be a problem with the IR code we are generating, since it should always be fully "type correct" at the IR level.